dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,346 @@
+        // to only keep & restore the names, associating them with IDs.
+        // It just wouldn't allow us to separately enable/disable persisting
+        // something else inside columns.
+        if (persisted_props.indexOf(propName) !== -1) {
+            const storage = stores[persistence_type];
+            const transform = getTransform(element, propName);
+
+            const newValKey = getNewValKey(id, propName);
+            const persistIdKey = getPersistIdKey(newValKey);
+            const setOriginalAndId = () => {
+                storage.setItem(
+                    getOriginalValKey(newValKey),
+                    transform.extract(props[propName])
+                );
+                storage.setItem(persistIdKey, persistence);
+            };
+            if (
+                !storage.hasItem(newValKey) ||
+                storage.getItem(persistIdKey) !== persistence
+            ) {
+                setOriginalAndId();
+            }
+            storage.setItem(newValKey, transform.extract(newPropVal));
+        }
+    }, newProps);
+}
+
+function clearUIEdit(id, persistence_type, propName) {
+    const storage = stores[persistence_type];
+    const newValKey = getNewValKey(id, propName);
+
+    if (storage.hasItem(newValKey)) {
+        storage.removeItem(newValKey);
+        storage.removeItem(getOriginalValKey(newValKey));
+        storage.removeItem(getPersistIdKey(newValKey));
+    }
+}
+
+/*
+ * Used for entire layouts (on load) or partial layouts (from children
+ * callbacks) to apply previously-stored UI edits to components
+ */
+export function applyPersistence(layout) {
+    if (type(layout) !== 'Object' || !layout.props) {
+        return layout;
+    }
+
+    return persistenceMods(layout, layout, []);
+}
+
+function persistenceMods(layout, component, path) {
+    const {
+        id,
+        props,
+        element,
+        persistence,
+        persisted_props,
+        persistence_type,
+    } = getProps(component);
+
+    let layoutOut = layout;
+    if (persistence) {
+        const storage = stores[persistence_type];
+        forEach(propName => {
+            const newValKey = getNewValKey(id, propName);
+            const storedPersistID = storage.getItem(getPersistIdKey(newValKey));
+            const transform = getTransform(element, propName);
+            if (storedPersistID) {
+                if (
+                    storedPersistID === persistence &&
+                    equals(
+                        storage.getItem(getOriginalValKey(newValKey)),
+                        transform.extract(props[propName])
+                    )
+                ) {
+                    layoutOut = set(
+                        lensPath(path.concat('props', propName)),
+                        transform.apply(
+                            storage.getItem(newValKey),
+                            props[propName]
+                        ),
+                        layoutOut
+                    );
```

Comment:
Concern here: If a component gets re-renderered and a prop has persisted changes, will the same value / object instance be returned as long as (baseValue, persisted) are the same?

e.g. If the table's column names are persisted and this returns a new array of columns on each render, the table's caching will bust on each render and degrade performance significantly.

Reply:
I don't *think* this is an issue - `applyPersistence` is not being called as part of the render loop, only on new data coming with the initial layout or from a callback. And that data (aside from clientside callbacks) all went through JSON anyway.

Reply:
Ok. Only re-evaluating If the value changed.

Reply:
Call me paranoid, tested it out with the table. As you said, only called on prop update.